students/elmar_m/lesson05/mailroom.py

```diff
@@ -0,0 +1,181 @@
+    # add new donation to donor and print letter of thanks.
+    Start submenu
+    '''
+    menu(sub_p, sub_d)
+
+
+def list_donors():
+    '''
+    Print list of donors
+    '''
+    print('\n'.join(donors))
+    return '\n'.join(donors)
+    
+
+def add():
+    '''
+    Add new donation and / or donor
+    '''
+    while True: 
+        dname = input('>> Please give donor name: ')
+        if not dname.isalpha():
+            print('>> only alphabetical characters in donor name allowed, please try again')
+            continue
+        else:
+            break
+    if dname in donors:
+        print('>>', dname, 'already in list')
+        add_amount(dname)
+    else:
+        print('>>', dname, 'not in list, adding it ')
+        add_amount(dname)
+
+
+def add_amount(donor):
+    '''
+    Add donation amount
+        
+        ARGS:
+    donor: name of donor
+    '''
+    while True:
+        amount = input('>> please add current donation (int or float, use \'.\' as decimal separator):\n>> ')
+        # make sure given donation values are positive: 
+        if '-' in amount:
+            print('>> only positive integers or floats allowed.')
+            continue
+        # Throw exception and ask again, if given value is not int or float:
+        try:
+            donors[donor].append(float(amount))
+            break
+        except ValueError:
+            print('>> only numerical values allowed, please try again')
+    print(amount, 'added to donation list of', donor, ', thank you')     
+
+
+def report():
+    '''
+    Print a formatted list of current donors and donations
+    '''
+    # get the highest number of digits to create formatstring accordingly:
+    sumlist = []
+    for i in donors.values():
+        sumlist.append(sum(i))
+    maxn = len(str(max(sumlist)))
+    maxn += 2
+    fstring = '{:<20} ' + '|' + '{:>' + str(maxn) + '} ' + '|' + '{:>9}' + '|' + '{:>20}' 
+    print(fstring.format('Donor Name', 'Total', 'Num Gifts', 'Average Gift'))  
+    print('-' * (maxn + 54)) 
+    for i in donors:
+        print(fstring.format(i, sum(donors[i]), len(donors[i]),  sum(donors[i])//len(donors[i]) )) 
+
+
+def efunc():
+    '''
+    Exit menu
+    '''
+    print('exiting.\n')
+    return 'exiting'
+
+
+main_d = {
+    # Dispatcher dictionary main menu
+    't' : thankyou,
+    'r' : report,
+    'q' : efunc,
+    }
+
+
+sub_d= {
+    #Dispatcher dictionary submenu
+    'l' : list_donors,
+    'a' : add,
+    'x' : efunc,
+    'm' : mail,
+    }
+
+
+def menu(p, d):
+    '''
+    Display menu to user. 
+    This function uses a "dispatcher dictionary".
+
+        ARGS:
+    p:  prompt which is shown to the user
+    d:  the dispatcher dictionary which holds the menu options
+    '''
+    try:
+        while True:
+            response = input(p)
+            if d[response]() == 'exiting':
+                break
+    except KeyError:
+        print('sorry, unknown option:', response)
+        print(p)
+        menu(p, d)
```

Comment:
you should put your KeyError around line 172, and inside of the except just the print statement so that the while True can continue to run